agentrs/src/ws/client.rs

```diff
@@ -0,0 +1,183 @@
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::sync::RwLock;
+use tracing::{debug, error, info};
+use tungstenite::{client::IntoClientRequest, handshake::client::Request};
+
+use futures::{SinkExt, StreamExt};
+use tokio::sync::Mutex;
+use tokio_tungstenite::{connect_async, tungstenite::protocol::Message};
+
+#[derive(Clone)]
+pub struct WebSocket {
+    pub config_manager: conf::ConfigHandleManager,
+    //pub gateway_url: String,
+    pub request: Request,
+    pub reconnect_interval: Duration,
+    pub max_reconnection_attempts: usize,
+}
+
+#[async_trait]
+impl Task for WebSocket {
+    type Output = anyhow::Result<()>;
+
+    const NAME: &'static str = "agent listener";
+
+    async fn run(self, mut shutdown_signal: ShutdownSignal) -> Self::Output {
+        tokio::select! {
+            result = self.run_with_reconnect() => result,
+            _ = shutdown_signal.wait() => Ok(()),
+        }
+    }
+}
+
+impl WebSocket {
+    pub fn new() -> anyhow::Result<Self> {
+        let config_manager =
+            conf::ConfigHandleManager::init().context("Failed to init config manager")?;
+
+        let gateway_url = std::env::var("GATEWAY_URL")
+            .unwrap_or_else(|_| "ws://localhost:8009/api/ws".to_string());
+        let mut request = gateway_url.into_client_request().unwrap();
+
+        // Insert a custom header
+        let token = config_manager.conf.token.clone().unwrap();
+        let token: &'static str = Box::leak(token.into_boxed_str());
+        request
+            .headers_mut()
+            .insert("HOOP_KEY", HeaderValue::from_static(token));
+
+        Ok(WebSocket {
+            request: request,
+            config_manager: config_manager,
+            reconnect_interval: Duration::from_secs(5),
+            max_reconnection_attempts: 10,
+        })
+    }
+
+    async fn run_with_reconnect(self) -> anyhow::Result<()> {
+        let mut attempts = 0;
+        loop {
+            match self.clone().run().await {
+                Ok(_) => {
+                    debug!("> WebSocket connection closed gracefully");
+                    return Ok(());
+                }
+                Err(e) if e.to_string().contains("connection closed") => {
+                    debug!("> WebSocket connection closed by server");
+                    return Ok(());
+                }
+                Err(e) if attempts >= self.max_reconnection_attempts => {
+                    debug!("> Max reconnection attempts reached, giving up: {}", e);
+                    return Err(e);
+                }
+                Err(e) => {
+                    attempts += 1;
+                    error!(
+                        "> Connection failed (attempt {}/{}): {}",
+                        attempts, self.max_reconnection_attempts, e
+                    );
+
+                    tokio::time::sleep(self.reconnect_interval).await;
+                    continue;
+                }
+            }
+        }
+    }
```

Comment:
The current agent implementation keeps trying to establish a connection until a connection is established with the gateway. It will only hang if there's an issue parsing the HOOP_KEY. It uses a backoff mechanism to avoid sending overloading the server

Not a blocker though

Reply:
Ok, so I will do the same here.